File: frontend/src/services/centroCustoService.ts
import api from './api';
import type { CentroCusto } from '../types';

interface CentroCustoFormData {
  descricao: string;
  linhaNegocioId: number;
}

const centroCustoService = {
  listarTodos: async (): Promise<CentroCusto[]> => {
    const response = await api.get('/centros-custo');
    return response.data;
  },

  buscarPorId: async (id: number): Promise<CentroCusto> => {
    const response = await api.get(`/centros-custo/${id}`);
    return response.data;
  },

  listarPorLinhaNegocio: async (linhaNegocioId: number): Promise<CentroCusto[]> => {
    const response = await api.get(`/centros-custo/linha-negocio/${linhaNegocioId}`);
    return response.data;
  },

  cadastrar: async (data: CentroCustoFormData): Promise<CentroCusto> => {
    const response = await api.post('/centros-custo', data);
    return response.data;
  },

  atualizar: async (id: number, data: CentroCustoFormData): Promise<CentroCusto> => {
    const response = await api.put(`/centros-custo/${id}`, data);
    return response.data;
  },

  remover: async (id: number): Promise<void> => {
    await api.delete(`/centros-custo/${id}`);
  }
};

export { centroCustoService }; 

File: backend/src/main/java/br/com/techne/sistemafolha/cadastros/application/CentroCustoService.java
package br.com.techne.sistemafolha.cadastros.application;

import br.com.techne.sistemafolha.cadastros.api.CentroCustoDTO;
import br.com.techne.sistemafolha.cadastros.domain.CentroCustoNotFoundException;
import br.com.techne.sistemafolha.cadastros.domain.LinhaNegocioNotFoundException;
import br.com.techne.sistemafolha.cadastros.domain.CentroCusto;
import br.com.techne.sistemafolha.cadastros.domain.LinhaNegocio;
import br.com.techne.sistemafolha.cadastros.infrastructure.CentroCustoRepository;
import br.com.techne.sistemafolha.cadastros.infrastructure.LinhaNegocioRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CentroCustoService {
    private final CentroCustoRepository centroCustoRepository;
    private final LinhaNegocioRepository linhaNegocioRepository;

    public List<CentroCustoDTO> listarTodas() {
        return centroCustoRepository.findByAtivoTrue().stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    public List<CentroCustoDTO> listarPorLinhaNegocio(Long linhaNegocioId) {
        return centroCustoRepository.findByLinhaNegocioIdAndAtivoTrue(linhaNegocioId).stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    public CentroCustoDTO buscarPorId(Long id) {
        return centroCustoRepository.findById(id)
                .filter(cc -> cc.getAtivo())
                .map(this::toDTO)
                .orElseThrow(() -> new CentroCustoNotFoundException(id));
    }

    @Transactional
    public CentroCustoDTO cadastrar(CentroCustoDTO dto) {
        LinhaNegocio linhaNegocio = linhaNegocioRepository.findById(dto.linhaNegocioId())
                .filter(ln -> ln.getAtivo())
                .orElseThrow(() -> new LinhaNegocioNotFoundException(dto.linhaNegocioId()));

        CentroCusto centroCusto = toEntity(dto);
        centroCusto.setLinhaNegocio(linhaNegocio);
        return toDTO(centroCustoRepository.save(centroCusto));
    }

    @Transactional
    public CentroCustoDTO atualizar(Long id, CentroCustoDTO dto) {
        CentroCusto centroCusto = centroCustoRepository.findById(id)
                .filter(cc -> cc.getAtivo())
                .orElseThrow(() -> new CentroCustoNotFoundException(id));

        LinhaNegocio linhaNegocio = linhaNegocioRepository.findById(dto.linhaNegocioId())
                .filter(ln -> ln.getAtivo())
                .orElseThrow(() -> new LinhaNegocioNotFoundException(dto.linhaNegocioId()));

        centroCusto.setDescricao(dto.descricao());
        centroCusto.setLinhaNegocio(linhaNegocio);
        return toDTO(centroCustoRepository.save(centroCusto));
    }

    @Transactional
    public void remover(Long id) {
        CentroCusto centroCusto = centroCustoRepository.findById(id)
                .filter(cc -> cc.getAtivo())
                .orElseThrow(() -> new CentroCustoNotFoundException(id));
        centroCusto.setAtivo(false);
        centroCustoRepository.save(centroCusto);
    }

    private CentroCustoDTO toDTO(CentroCusto centroCusto) {
        return new CentroCustoDTO(
            centroCusto.getId(),
            centroCusto.getDescricao(),
            centroCusto.getAtivo(),
            centroCusto.getLinhaNegocio().getId()
        );
    }

    private CentroCusto toEntity(CentroCustoDTO dto) {
        CentroCusto centroCusto = new CentroCusto();
        centroCusto.setDescricao(dto.descricao());
        centroCusto.setAtivo(true);
        return centroCusto;
    }
} 